# 2. Data Cleaning

In [80]:
import pandas as pd
import numpy as np
import datetime as dt


In [81]:
path = 'C:\\Users\\Arnold\\OneDrive\\Desktop\\Knowledge Discovery and Data Mining\\Car Regression\\car_prices_regression\\data\\raw'

test_data = pd.read_csv(path + "\\test.csv")
train_data = pd.read_csv(path +"\\train.csv")

In [82]:
#Combining the 2 data sets to perform data visualization on them

x_train = train_data.drop("price", axis=1)
y_train = train_data["price"]
    
combined_data_set = pd.concat([x_train,test_data],axis=0, ignore_index=True)

y_train.to_csv("C:\\Users\\Arnold\\OneDrive\\Desktop\\Knowledge Discovery and Data Mining\\Car Regression\\car_prices_regression\\data\\interim\\price.csv")

In [83]:
combined_data_set.head()

id          brand              model  model_year  milage      fuel_type  \
0   0           MINI      Cooper S Base        2007  213000       Gasoline   
1   1        Lincoln              LS V8        2002  143250       Gasoline   
2   2      Chevrolet  Silverado 2500 LT        2002  136731  E85 Flex Fuel   
3   3        Genesis   G90 5.0 Ultimate        2017   19500       Gasoline   
4   4  Mercedes-Benz        Metris Base        2021    7388       Gasoline   

                                              engine  \
0       172.0HP 1.6L 4 Cylinder Engine Gasoline Fuel   
1       252.0HP 3.9L 8 Cylinder Engine Gasoline Fuel   
2  320.0HP 5.3L 8 Cylinder Engine Flex Fuel Capab...   
3       420.0HP 5.0L 8 Cylinder Engine Gasoline Fuel   
4       208.0HP 2.0L 4 Cylinder Engine Gasoline Fuel   

                     transmission ext_col int_col  \
0                             A/T  Yellow    Gray   
1                             A/T  Silver   Beige   
2                             A/T    Blue    Gray   
3  Transmission w/Dual Shift Mode   Black   Black   
4                     7-Speed A/T   Black   Beige   

                                 accident clean_title  
0                           None reported         Yes  
1  At least 1 accident or damage reported         Yes  
2                           None reported         Yes  
3                           None reported         Yes  
4                           None reported         Yes

In [84]:
## Checking the null values in fuel_type column

fuel_null = combined_data_set.loc[combined_data_set['fuel_type'] == "–"]
fuel_null

id      brand                     model  model_year  milage  \
532        532      Dodge  Challenger R/T Scat Pack        2021   35000   
1155      1155       Ford        Mustang GT Premium        2017  143600   
1307      1307     Toyota         Land Cruiser Base        1994  138033   
1362      1362       Ford  Mustang EcoBoost Premium        2016   43000   
1741      1741  Chevrolet                Protege DX        1993   94200   
...        ...        ...                       ...         ...     ...   
313202  313202     Nissan                300ZX Base        1993  168000   
313318  313318       Ford  Mustang EcoBoost Premium        2016  111300   
313412  313412     Toyota         Land Cruiser Base        1996   58500   
313634  313634  Chevrolet             1500 Cheyenne        1994  158000   
313877  313877      Dodge        Challenger SRT 392        2016   62000   

       fuel_type engine transmission ext_col int_col  \
532            –      –  8-Speed A/T    Blue    Gray   
1155           –      –  6-Speed M/T   White   Black   
1307           –      –          A/T   Green   Beige   
1362           –      –          A/T    Blue   Black   
1741           –      –          A/T     Red    Gray   
...          ...    ...          ...     ...     ...   
313202         –      –  5-Speed M/T     Red   Black   
313318         –      –          A/T  Orange   Black   
313412         –      –          A/T   White   Black   
313634         –      –          A/T   White   Brown   
313877         –      –  8-Speed A/T     Red   Black   

                                      accident clean_title  
532                              None reported         Yes  
1155    At least 1 accident or damage reported         Yes  
1307                             None reported         Yes  
1362                             None reported         Yes  
1741                             None reported         Yes  
...                                        ...         ...  
313202                           None reported         Yes  
313318                           None reported         Yes  
313412                           None reported         Yes  
313634  At least 1 accident or damage reported         Yes  
313877                           None reported         Yes  

[1319 rows x 12 columns]

In [85]:

(len(fuel_null.id)/len(combined_data_set.id)) * 100



0.41976558049538065

We could drop the rows with the null values in the fuel_type column since it is a marginal proportion of all vehicles

In [86]:
combined_data_set = combined_data_set[~(combined_data_set['fuel_type'] == "–")]

In [87]:
not_supported_fuel  = combined_data_set.loc[combined_data_set['fuel_type'] == "not supported"]
not_supported_fuel

id      brand                           model  model_year  milage  \
4823      4823  Chevrolet                     Suburban LT        2011  109646   
18000    18000    Porsche                      Camaro 2SS        2015   67000   
19182    19182     Nissan                   Mirai Limited        2000  131000   
27199    27199     Subaru            Outback 2.5i Limited        2017  116000   
33030    33030     Nissan                   Mirai Limited        2009  137333   
60717    60717  Chevrolet              Silverado 1500 RST        2020   19200   
74096    74096     Toyota                   Mirai Limited        2021   79876   
86897    86897      Lexus                  IS 350 F Sport        2022    9500   
94425    94425    Porsche                Cayenne Turbo GT        2018   52000   
127066  127066     Toyota                   Mirai Limited        2016   58664   
128917  128917      Mazda                 MX-5 Miata Base        1999  120000   
135865  135865      Lexus                   Mirai Limited        2014  123331   
140206  140206      Lexus                     RX 330 Base        2006  110000   
171120  171120     Hummer                         H2 Base        2013  152068   
171659  171659     Toyota                   Mirai Limited        2021   44000   
200729  200729     Toyota                      Mirai Base        2011   91491   
209488  209488       Alfa          Romeo Stelvio Ti Sport        2018   23600   
229103  229103       Audi                 A3 2.0T Premium        2015  205500   
249628  249628     Toyota                      Mirai Base        2018   18000   
256193  256193     Nissan                   Mirai Limited        2016   49688   
259042  259042     Toyota                      Mirai Base        2015   46873   
262931  262931  Chevrolet         Corvette Stingray w/2LT        2018    5400   
264225  264225       Land  Rover Range Rover Supercharged        2014  134000   
266935  266935       Ford                   Focus ST Base        2018   49500   
273147  273147     Toyota                   Mirai Limited        2021   20500   
279874  279874     Toyota                   Mirai Limited        2011  107000   
285151  285151   Cadillac              Escalade ESV Sport        2018   32000   
308236  308236     Subaru                    WRX STI Base        2014   76500   
310455  310455     Toyota                      Mirai Base        2016   74000   

            fuel_type                                             engine  \
4823    not supported  320.0HP 5.3L 8 Cylinder Engine Flex Fuel Capab...   
18000   not supported       240.0HP 2.0L 4 Cylinder Engine Gasoline Fuel   
19182   not supported               182.0HP Electric Motor Hydrogen Fuel   
27199   not supported       175.0HP 2.5L 4 Cylinder Engine Gasoline Fuel   
33030   not supported               182.0HP Electric Motor Hydrogen Fuel   
60717   not supported       355.0HP 5.3L 8 Cylinder Engine Gasoline Fuel   
74096   not supported               182.0HP Electric Motor Hydrogen Fuel   
86897   not supported      311.0HP 3.5L V6 Cylinder Engine Gasoline Fuel   
94425   not supported       550.0HP 4.8L 8 Cylinder Engine Gasoline Fuel   
127066  not supported               151.0HP Electric Motor Hydrogen Fuel   
128917  not supported       140.0HP 1.8L 4 Cylinder Engine Gasoline Fuel   
135865  not supported               182.0HP Electric Motor Hydrogen Fuel   
140206  not supported      230.0HP 3.3L V6 Cylinder Engine Gasoline Fuel   
171120  not supported       316.0HP 6.0L 8 Cylinder Engine Gasoline Fuel   
171659  not supported               182.0HP Electric Motor Hydrogen Fuel   
200729  not supported               151.0HP Electric Motor Hydrogen Fuel   
209488  not supported       280.0HP 2.0L 4 Cylinder Engine Gasoline Fuel   
229103  not supported       211.0HP 2.0L 4 Cylinder Engine Gasoline Fuel   
249628  not supported               151.0HP Electric Motor Hydrogen Fuel   
256193  not supported               151.0HP Electric Motor Hy

In [88]:

(len(not_supported_fuel.id)/len(combined_data_set.id)) * 100


0.009268018305934088

In [89]:
is_na_fuel_type = combined_data_set.loc[combined_data_set["fuel_type"].isna()]
is_na_fuel_type


id   brand                  model  model_year  milage fuel_type  \
11          11   Tesla          Model S P100D        2015   81500       NaN   
32          32   Tesla          Model S P100D        2018   30300       NaN   
51          51   Tesla     Model 3 Long Range        2020   38360       NaN   
52          52   Tesla            Model X 75D        2021   88000       NaN   
161        161   Tesla     Model S Long Range        2021    9500       NaN   
...        ...     ...                    ...         ...     ...       ...   
313955  313955  Rivian  R1S Adventure Package        2023    9000       NaN   
314063  314063   Tesla     Model 3 Long Range        2018   42094       NaN   
314201  314201   Tesla           Model S 100D        2018   58000       NaN   
314210  314210   Tesla     Model 3 Long Range        2019   69548       NaN   
314217  314217   Tesla             Model S 85        2014   67647       NaN   

                                             engine transmission ext_col  \
11              Electric Motor Electric Fuel System  1-Speed A/T    Gray   
32      518.0HP Electric Motor Electric Fuel System          A/T   White   
51                            Dual Motor - Standard    Automatic   White   
52      518.0HP Electric Motor Electric Fuel System  1-Speed A/T   Black   
161     670.0HP Electric Motor Electric Fuel System          A/T   White   
...                                             ...          ...     ...   
313955  835.0HP Electric Motor Electric Fuel System          A/T     Red   
314063  271.0HP Electric Motor Electric Fuel System          A/T   Black   
314201  518.0HP Electric Motor Electric Fuel System  1-Speed A/T    Gray   
314210          Electric Motor Electric Fuel System  1-Speed A/T     Red   
314217  362.0HP Electric Motor Electric Fuel System  1-Speed A/T    Blue   

       int_col                                accident clean_title  
11       White                           None reported         Yes  
32       White                           None reported         Yes  
51       Black                           None reported         NaN  
52       Black                           None reported         Yes  
161      Beige                           None reported         Yes  
...        ...                                     ...         ...  
313955   White                           None reported         Yes  
314063   Black                           None reported         Yes  
314201   Black                           None reported         Yes  
314210   Black                           None reported         Yes  
314217   Black  At least 1 accident or damage reported         Yes  

[8466 rows x 12 columns]

It appears most of these vehicles are electric 

In [90]:
(len(is_na_fuel_type.id)/len(combined_data_set.id)) * 100


2.7056221716564828

In [91]:
## Checking if all vehicles are electric

is_na_fuel_type[~is_na_fuel_type['engine'].str.contains("electric|Dual Motor|battery|kw" , case = False)]


id          brand                    model  model_year  milage  \
3286      3286          smart              ForTwo Pure        2008  156875   
6794      6794      Chevrolet  Corvette Stingray w/2LT        2023   10900   
8879      8879           Audi   Q5 S line Premium Plus        2016   35315   
8908      8908            BMW                    M3 CS        2022   13777   
9522      9522            Kia                 EV6 Wind        2022    4514   
...        ...            ...                      ...         ...     ...   
308824  308824          Tesla              Model S 75D        2017   51808   
310616  310616  Mercedes-Benz     C-Class C 300 4MATIC        2018   64300   
312225  312225        Porsche           718 Cayman GT4        2021   23600   
312319  312319           Ford   F-150 Lightning LARIAT        2020   18385   
312982  312982           Audi          S5 3.0T Premium        2023     654   

       fuel_type                                             engine  \
3286         NaN        70.0HP 1.0L 3 Cylinder Engine Gasoline Fuel   
6794         NaN       490.0HP 6.2L 8 Cylinder Engine Gasoline Fuel   
8879         NaN                                     2.0 Liter TFSI   
8908         NaN  453.0HP 3.0L Straight 6 Cylinder Engine Gasoli...   
9522         NaN                         1.6L I4 16V GDI DOHC Turbo   
...          ...                                                ...   
308824       NaN                               5.0L V8 32V PDI DOHC   
310616       NaN       261.0HP 2.0L 4 Cylinder Engine Gasoline Fuel   
312225       NaN  414.0HP 4.0L Flat 6 Cylinder Engine Gasoline Fuel   
312319       NaN                               5.0L V8 32V PDI DOHC   
312982       NaN      349.0HP 3.0L V6 Cylinder Engine Gasoline Fuel   

             transmission                ext_col      int_col  \
3286          2-Speed A/T                  Black         Gray   
6794          8-Speed A/T                   Blue        Brown   
8879            Automatic           Daytona Gray        Black   
8908            Automatic                   Gray        Black   
9522            Automatic  Pacific Blue Metallic  Satin Black   
...                   ...                    ...          ...   
308824  1-Speed Automatic                  Black        Black   
310616                A/T                  Black            –   
312225        6-Speed M/T                    Red        Black   
312319  1-Speed Automatic           Caspian Blue         Gray   
312982        1-Speed A/T                   Gray        Black   

                                      accident clean_title  
3286    At least 1 accident or damage reported         Yes  
6794                             None reported         Yes  
8879    At least 1 accident or damage reported         NaN  
8908                             None reported         Yes  
9522                                       NaN         NaN  
...                                        ...         ...  
308824                           None reported         NaN  
310616                           None reported         Yes  
312225                           None reported         Yes  
312319                                     NaN         NaN  
312982                           None reported         Yes  

[162 rows x 12 columns]

We could add a column to confirm whether a car is electric or not. It appears that some vehicles have been modded e.g. There is a Tesla with a 5 liter engine. However, it is difficult to ascertain if all cars that fall under this category has been modded. For example the Acura MDX 3.5L is not electric. Additionally, there are only 162 rows of these vehicles that do not fit in the electric category or fuel category without further cleaning.

In [92]:
electric_cars_list  = "electric|Dual Motor|battery|kw"

combined_data_set.loc[:,"electric"] = np.where(combined_data_set["engine"].str.contains(electric_cars_list,case = False, na= False),1,0)

In [93]:
##Checking the accident status

combined_data_set.accident.unique()



array(['None reported', 'At least 1 accident or damage reported', nan],
      dtype=object)

In [94]:
combined_data_set.loc[combined_data_set["accident"].isna()].head()

id          brand                 model  model_year  milage fuel_type  \
87    87            RAM         2500 Big Horn        2012   95400    Diesel   
134  134           Jeep       Gladiator Sport        2021   61341  Gasoline   
164  164          Acura  RDX w/A-Spec Package        2023   39517  Gasoline   
333  333  Mercedes-Benz      SLK-Class SLK320        2001  170000  Gasoline   
361  361            RAM              2500 SLT        2016   69600  Gasoline   

                                                engine       transmission  \
87   350.0HP 6.7L Straight 6 Cylinder Engine Diesel...        6-Speed A/T   
134                              3.6L V6 24V MPFI DOHC  8-Speed Automatic   
164                               2.0 Liter DOHC Turbo          Automatic   
333      215.0HP 3.2L V6 Cylinder Engine Gasoline Fuel        5-Speed A/T   
361       410.0HP 6.4L 8 Cylinder Engine Gasoline Fuel        6-Speed A/T   

                       ext_col int_col accident clean_title  electric  
87                       Black   Black      NaN         NaN         0  
134  Firecracker Red Clearcoat   Black      NaN         NaN         0  
164                       Gray   Black      NaN         NaN         0  
333                      Black   Beige      NaN         NaN         0  
361                       Blue    Gray      NaN         NaN         0

In [95]:
combined_data_set.loc[combined_data_set["accident"].isna()].tail()

id  brand                  model  model_year  milage fuel_type  \
313559  313559   Jeep     Gladiator Overland        2020   49814  Gasoline   
313883  313883    BMW           X6 xDrive50i        2019   57918  Gasoline   
313928  313928   Ford               F-250 XL        2022   38012  Gasoline   
313957  313957   Audi       S5 3.0T Prestige        2019   67015  Gasoline   
314025  314025  Dodge  Ram 3500 Quad Cab DRW        2006  110500    Diesel   

                                                   engine       transmission  \
313559                              3.6L V6 24V MPFI DOHC  8-Speed Automatic   
313883       445.0HP 4.4L 8 Cylinder Engine Gasoline Fuel        8-Speed A/T   
313928                               5.7L V8 16V MPFI OHV          Automatic   
313957                                               3.0L  1-Speed Automatic   
314025  400.0HP 6.7L Straight 6 Cylinder Engine Diesel...        6-Speed A/T   

                          ext_col int_col accident clean_title  electric  
313559  Firecracker Red Clearcoat   Black      NaN         NaN         0  
313883                      White   Beige      NaN         NaN         0  
313928    Delmonico Red Pearlcoat   Black      NaN         NaN         0  
313957  Daytona Gray Pearl Effect   Black      NaN         NaN         0  
314025                        Red    Gray      NaN         NaN         0

In [96]:
len(combined_data_set.loc[combined_data_set["accident"].isna()])

4062

In [97]:
len(combined_data_set.loc[combined_data_set["accident"].isna() & combined_data_set["clean_title"].isna()])


4039

From our initial analysis, we can see that most cars missing the accident value also are missing the clean title value. This is because it might not be possible to check whether the vehicle has been involved in an accident using the VIN number. Since approximately 12% of the data is slightly significant, we could fill the NA values with unknown. Additionally, more information may be available upon further analysis

In [98]:
combined_data_set.loc[:,"accident"] = combined_data_set["accident"].fillna("unknown")
combined_data_set.loc[:,"clean_title"] = combined_data_set["clean_title"].fillna("unknown")

In [99]:
## Creating a feature for age
max_year = dt.date(2021,1,1).year
combined_data_set["model_year"] = pd.to_datetime(combined_data_set["model_year"], format = "%Y").dt.year
combined_data_set["age"] = max_year - combined_data_set.model_year


In [100]:
combined_data_set.columns

Index(['id', 'brand', 'model', 'model_year', 'milage', 'fuel_type', 'engine',
       'transmission', 'ext_col', 'int_col', 'accident', 'clean_title',
       'electric', 'age'],
      dtype='object')

In [101]:
## Creating a feature of brand popularity

brand_popularity = combined_data_set.brand.value_counts()
brand_popularity_counts = combined_data_set["brand"].map(brand_popularity)
popularity_bins = [0,5000,10000,40000]
popularity_labels = ["low", "medium", "high"]
combined_data_set["brand_popularity"]  = pd.cut(brand_popularity_counts, bins = popularity_bins, labels=popularity_labels, include_lowest=True)

In [102]:
## Creating a feature for miles per year

combined_data_set["miles_per_year"] = combined_data_set.milage / combined_data_set.age


In [103]:
## Creating a feature for color popularity

color_counts = combined_data_set['ext_col'].value_counts()
color_bins = [0,2000,10000,90000]
color_labels = ["low", "medium", "high"]
color_popularity = combined_data_set.ext_col.map(color_counts)

combined_data_set["color_popularity"] = pd.cut(color_popularity,bins = color_bins,labels=color_labels,include_lowest=True)

In [104]:
# Dropping the ID columns

combined_data_set = combined_data_set.drop("id", axis = 1)